File: app/src/main/java/com/attendance/attendresults/GetAttendResultsSubscriber.java
package com.attendance.attendresults;

import com.attendance.AttendanceApplication;
import com.attendance.R;
import com.attendance.data.AttendResult;

import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.List;

import rx.Subscriber;

/**
 * Created by peiqin on 3/5/2017.
 */

public class GetAttendResultsSubscriber extends Subscriber<AttendResult> {

    private AttendResultsContract.View view;
    private List<AttendResult> list;

    public GetAttendResultsSubscriber(AttendResultsContract.View view) {
        this.view = view;
    }

    @Override
    public void onStart() {
        list = new ArrayList<>();
    }

    @Override
    public void onCompleted() {
        view.getAttendResultsSuccess(list);
        view.showTip(AttendanceApplication.getContext().getString(R.string.get_success));
        if (!this.isUnsubscribed()) {
            this.unsubscribe();
        }
    }

    @Override
    public void onError(Throwable e) {

        if (e instanceof SocketTimeoutException || e instanceof ConnectException) {
            view.showTip(AttendanceApplication.getContext().getString(R.string.network_corrupt));
        } else {
            view.showTip("error:" + e.getMessage());
        }

        if (!this.isUnsubscribed()) {
            this.unsubscribe();
        }

    }

    @Override
    public void onNext(AttendResult resultBean) {

        list.add(resultBean);

    }

}

File: app/src/main/kotlin/com.attendance/attendresults/GetAttendResultsSubscriber.kt
package com.attendance.attendresults

import android.app.Application
import com.attendance.AttendanceApplication
import com.attendance.R
import com.attendance.data.AttendResult

import java.net.ConnectException
import java.net.SocketTimeoutException
import java.util.ArrayList

import rx.Subscriber

/**
 * Created by peiqin on 3/5/2017.
 */

class GetAttendResultsSubscriber(private val view: AttendResultsContract.View) : Subscriber<AttendResult>() {
    private var list: MutableList<AttendResult> = ArrayList<AttendResult>()

    override fun onStart() {
    }

    override fun onCompleted() {
        view.getAttendResultsSuccess(list)
        view.showTip(AttendanceApplication.context.getString(R.string.get_success))
        if (!this.isUnsubscribed) {
            this.unsubscribe()
        }
    }

    override fun onError(e: Throwable) {
        if (e is SocketTimeoutException || e is ConnectException) {
            view.showTip(AttendanceApplication.context.getString(R.string.network_corrupt))
        } else {
            view.showTip("error:" + e.message)
        }

        if (!this.isUnsubscribed) {
            this.unsubscribe()
        }
    }

    override fun onNext(attendResult: AttendResult) {
        list.add(attendResult)
    }

}
